safet/optional.hpp

```diff
@@ -161,182 +162,189 @@ class optional {
         }
     }
 
-    template <typename Functor>
-    decltype(auto) if_set(Functor&& f) &
+   template <typename Functor, typename... AdditionalArgs>
+    decltype(auto) if_set(Functor&& f, AdditionalArgs&&... additional_args) &
     {
-        static_assert(impl::invocable<Functor&&, T&>, "if_set functor must be invocable with T&");
+        static_assert(impl::invocable<Functor&&, T&, AdditionalArgs&&...>, "if_set functor must be invocable with T&");
 
-        if constexpr (impl::invocable_and_returns_something<Functor&&, T&>) {
-            return [&]() -> optional<std::invoke_result_t<Functor&&, T&>> {
+        if constexpr (impl::invocable_and_returns_something<Functor&&, T&, AdditionalArgs&&...>) {
+            return [&]() -> optional<std::invoke_result_t<Functor&&, T&, AdditionalArgs&&...>> {
                 if (m_engaged) {
-                    return std::forward<Functor>(f)(value());
+                    return std::forward<Functor>(f)(value(), std::forward<AdditionalArgs>(additional_args)...);
                 }
 
                 return std::nullopt;
             }();
         } else {
             if (m_engaged) {
-                std::forward<Functor>(f)(value());
+                std::forward<Functor>(f)(value(), std::forward<AdditionalArgs>(additional_args)...);
             }
 
             return *this;
         }
     }
 
-    template <typename Functor>
-    decltype(auto) if_set(Functor&& f) const&
+   template <typename Functor, typename... AdditionalArgs>
+    decltype(auto) if_set(Functor&& f, AdditionalArgs&&... additional_args) const&
     {
-        static_assert(impl::invocable<Functor&&, const T&>, "if_set functor on const optional must be invocable with const T&");
+        static_assert(impl::invocable<Functor&&, const T&, const AdditionalArgs&&...>, "if_set functor on const optional must be invocable with const T&");
 
-        if constexpr (impl::invocable_and_returns_something<Functor&&, const T&>) {
-            return [&]() -> optional<std::invoke_result_t<Functor&&, const T&>> {
+        if constexpr (impl::invocable_and_returns_something<Functor&&, const T&, const AdditionalArgs&&...>) {
+            return [&]() -> optional<std::invoke_result_t<Functor&&, const T&, const AdditionalArgs&&...>> {
```

Comment:
the consts here should not be applied to AdditionalArgs (only to the optional's value), constifying the forwarding references will break forwarding, a unit test where one of the additional arguments is a move-only type, like an std::unique_ptr, could be used to ensure forwarding works in more cases

Reply:
right, there are a few places right here where you're adding "const" to "AdditionalArgs", you don't want that (because it means AdditionalArgs can't be moved when appropriate). I didn't see it in any other functions, though.